src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -2,88 +2,129 @@
-                (uri) => new DocumentSynchronizingContext(virtualDocument, document.Version, _synchronizationTimeout, cancellationToken),
-                (uri, existingContext) =>
+            lock (DocumentContextLock)
+            {
+                if (!_virtualDocumentContexts.TryGetValue(virtualDocumentUri, out var documentContext))
                 {
-                    if (virtualDocument == existingContext.VirtualDocument &&
-                        document.Version == existingContext.ExpectedHostDocumentVersion)
-                    {
-                        // Already contain a synchronizing context that represents this request and it's in-process of being calculated.
-                        return existingContext;
-                    }
+                    return;
+                }
 
-                    // Cancel old request
-                    existingContext.SetSynchronized(false);
-                    return new DocumentSynchronizingContext(virtualDocument, document.Version, _synchronizationTimeout, cancellationToken);
-                });
+                if (!textBuffer.TryGetHostDocumentSyncVersion(out var hostDocumentVersion))
+                {
+                    return;
+                }
 
-            var result = await _joinableTaskFactory.RunAsync(() => synchronizingContext.OnSynchronizedAsync);
+                documentContext.UpdateSeenDocumentVersion(hostDocumentVersion);
 
-            _synchronizingContexts.TryRemove(virtualDocument.Uri, out _);
+                var synchronizingContext = documentContext.SynchronizingContext;
+                if (synchronizingContext == null)
+                {
+                    // No active synchronizing context for this document.
+                    return;
+                }
 
-            return result;
+                if (documentContext.SeenHostDocumentVersion == synchronizingContext.RequiredHostDocumentVersion)
+                {
+                    // The buffers host document version indicates that it's now "synchronized". If we were to re-request the VirtualDocumentSnapshot
+                    // from the LSPDocumentManager we would see a synchronized LSPDocument. Instead of re-requesting we're making the assumption that
+                    // whenever a VirtualDocumentSnapshot is updated it also updates its ITextBuffer sync version.
+                    synchronizingContext.SetSynchronized(true);
+                }
+                else if (documentContext.SeenHostDocumentVersion > synchronizingContext.RequiredHostDocumentVersion)
+                {
+                    // The LSP document version has surpassed what the projected document was expecting for a version. No longer able to synchronize.
+                    synchronizingContext.SetSynchronized(false);
+                }
+                else
+                {
+                    // Seen host document version is less than the required version, need to wait longer.
```

Comment:
To me "need to wait longer" implies some kind of active retry, but I take it in this instance that you mean "it will likely be updated on the next sync request"?

Reply:
Nope, the typing flow in our LSP editor is as we type our language server continually process' documents behind the scenes and updates open documents C# buffers on the client. Every time we "update a C# buffer" on the client we're updating the C# buffer for a given host document version. Aka typing:

"a"   "b"   "c"  to get "abc". Results in 3 host document versions ("a" 1, "ab" 2, "abc" 3). Given our language server doesn't guarantee it's going to provide C# updates for every host document version it's possible that we only get a C# buffer update for 1 and for 3 but not for 2. Basically our language server guarantees that it will "eventually" provide a C# update for the latest document but doesn't make any promise for inbetween states of that document.

This comment is referring to the case that we've typed "c" but haven't yet received the C# update behind the scenes the corresponds to the host document version 3. Meaning our "seen " version could still be at 1 but given we've been asked to synchronize against "3" we're guaranteed to at some point get C# for a host document >= 3, so we need to wait longer til we see that.